### DISCUSSION TITLE
How to handle service layer error

### DISCUSSION THREAD

**Discussion by dennissetiawan** [indeedeng/iwf#285] [Q&A]
If i have a workflow that waits for a signal then the workflow execute a service that returns error because of some business logic validation  (e.g. the business logic does not allow inserting price with amount 0 or the user is not authorized to access a data) . How do we handle this error and make the wf recover to the proper state(e.g. wait for another input) ? because it will do backoff retry with the same signal value forever.

**Comment by longquanzheng** [comment]
The default infinite backoff retry doesn’t mean that you just wait for the errors. Monitor should be setup to get alert on errors (either using Cadence/Temporal activity error metrics or the iwf application rest api metrics) . So you should be able to get alert and mitigate before workflow timeout. 

When you have confirmed the root cause, like the signal value is invalid, you can update the workflow code to catch the exception, and then skip the error to go back to wait for another valid signal. 

Note that unlike Temporal/Cadence, updating the workflow code like that will take effect immediately without needing to do any “versioning “

**Reply by dennissetiawan** [reply]
> update the workflow code to catch the exception, and then skip the error to go back to wait for another valid signal.

does this mean we should handle this by transitioning to the same state as the next state? ( i.e. stateA -> stateA ) so that it goes back to wait until and wait for the signal value?
this seems weird to me because that means every state would have this kind of loop to the same state

**Reply by longquanzheng** [reply]
If so very likely that there is something wrong in the overall design. Esp why would you send an invalid signal to a workflow. It’s okay if it’s a mistake/bug and then we update the code to fix it. But you shouldn’t use this from beginning. 

Another feature to consider is “RPC” which allow you to validate the input and return an error early before sending to the workflow.

**Reply by dennissetiawan** [reply]
i see , so that means i need to move out the code for expected FE error like authorization or some business logic handling outside the workflow. Thanks for answering ! 

Also when should we consider using RPC? because in my use case , i can just handle this in the http handler before going to the workflow

**Reply by longquanzheng** [reply]
Putting the validation logic in workflow RPC will be useful for many cases. The RPC will have access to workflow persistence for read write, and publish message to internal channel and even trigger new state execution. 

There is a section of signal channel vs RPC in readme. Generally, RPC is much more flexible and powerful than signal channel
